// @stryker-mutator/jest-runner/dist/src/jest-plugins/with-coverage-analysis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withCoverageAnalysis = void 0;
const util_1 = require("@stryker-mutator/util");
const JEST_CIRCUS_RUNNER = 'jest-circus/runner';
/**
 * Jest's defaults.
 * @see https://jestjs.io/docs/en/configuration
 */
const jestDefaults = Object.freeze({
    testRunner: 'jest-jasmine2',
    testEnvironment: 'jsdom',
});
const setupFilesForPerTestCoverageAnalysis = {
    'jest-jasmine2': require.resolve('./jasmine2-setup-coverage-analysis'),
};
const testEnvironmentOverrides = {
    node: require.resolve('./jest-environment-node'),
    jsdom: require.resolve('./jest-environment-jsdom'),
    'jsdom-sixteen': require.resolve('./jest-environment-jsdom-sixteen'),
};
function getName(requireId, optionalPrefix) {
    if (requireId.startsWith(optionalPrefix)) {
        return requireId.substr(optionalPrefix.length);
    }
    else {
        return requireId;
    }
}
function withCoverageAnalysis(jestConfig, coverageAnalysis) {
    // Override with Stryker specific test environment to capture coverage analysis
    if (coverageAnalysis === 'off') {
        return jestConfig;
    }
    else {
        const overrides = {};
        overrideEnvironment(jestConfig, overrides);
        if (coverageAnalysis === 'perTest') {
            setupFramework(jestConfig, overrides);
        }
        return Object.assign(Object.assign({}, jestConfig), overrides);
    }
}
exports.withCoverageAnalysis = withCoverageAnalysis;
function setupFramework(jestConfig, overrides) {
    var _a, _b;
    const setupFile = setupFilesForPerTestCoverageAnalysis[(_a = jestConfig.testRunner) !== null && _a !== void 0 ? _a : jestDefaults.testRunner];
    if (setupFile) {
        overrides.setupFilesAfterEnv = [setupFile, ...((_b = jestConfig.setupFilesAfterEnv) !== null && _b !== void 0 ? _b : [])];
    }
    else if (jestConfig.testRunner !== JEST_CIRCUS_RUNNER) {
        // 'jest-circus/runner' is supported, via handleTestEvent, see https://jestjs.io/docs/en/configuration#testenvironment-string
        throw new Error(`The @stryker-mutator/jest-runner doesn't support ${util_1.propertyPath('coverageAnalysis')} "perTest" with "jestConfig.testRunner": "${jestConfig.testRunner}". Please open an issue if you want support for this: https://github.com/stryker-mutator/stryker-js/issues`);
    }
}
function overrideEnvironment(jestConfig, overrides) {
    var _a;
    const jestEnvironment = getName((_a = jestConfig.testEnvironment) !== null && _a !== void 0 ? _a : jestDefaults.testEnvironment, 'jest-environment-');
    if (testEnvironmentOverrides[jestEnvironment]) {
        overrides.testEnvironment = testEnvironmentOverrides[jestEnvironment];
    }
}
